state-chain/pallets/cf-pools/src/lib.rs

```diff
@@ -987,13 +1002,49 @@ impl<T: Config> SwappingApi for Pallet<T> {
 		let (asset_pair, order) =
 			AssetPair::from_swap(from, to).ok_or(Error::<T>::PoolDoesNotExist)?;
 		Self::try_mutate_pool(asset_pair, |_asset_pair, pool| {
-			let (output_amount, remaining_amount) =
-				pool.pool_state.swap(order, input_amount.into(), None);
-			remaining_amount
-				.is_zero()
-				.then_some(())
-				.ok_or(Error::<T>::InsufficientLiquidity)?;
-			let output_amount = output_amount.try_into().map_err(|_| Error::<T>::OutputOverflow)?;
+			let output_amount = if input_amount == 0 {
+				0
+			} else {
+				let input_amount: Amount = input_amount.into();
+
+				let tick_before = pool
+					.pool_state
+					.current_price(order)
+					.ok_or(Error::<T>::InsufficientLiquidity)?
+					.2;
+				let (output_amount, _remaining_amount) =
+					pool.pool_state.swap(order, input_amount, None);
+				let tick_after = pool
+					.pool_state
+					.current_price(order)
+					.ok_or(Error::<T>::InsufficientLiquidity)?
+					.2;
+
+				let swap_tick = common::tick_at_sqrt_price(
+					PoolState::<(T::AccountId, OrderId)>::swap_sqrt_price(
+						order,
+						input_amount,
+						output_amount,
+					),
+				);
+				let bounded_swap_tick = if tick_after < tick_before {
+					core::cmp::min(core::cmp::max(tick_after, swap_tick), tick_before)
+				} else {
+					core::cmp::min(core::cmp::max(tick_before, swap_tick), tick_after)
+				};
```

Comment:
Are these checks just there to be defensive? How could the swap tick not be bounded by the start and end ticks?

Reply:
Rounding, can cause that. Particularly with small values, i.e. 1 outputing 0, would slip the price infinitely, causing these checks to always fail, this is a defence again that.

This is not just a problem with directly around zero, like 10 outputing 9, slips the price 10%, but real slippage may only be 1%, but we only output 9 becuase of rounding.